examples/hikari_tanjun_basic_queue/client.py

```diff
@@ -0,0 +1,106 @@
+import os
+import typing
+
+import hikari
+import tanjun
+import lavasnek_rs
+
+
+PREFIX = ","
+TOKEN = os.environ["TOKEN"]
+
+
+class EventHandler:
+    """Handles events from the Lavalink server."""
+
+    async def track_start(self, _: lavasnek_rs.Lavalink, event: lavasnek_rs.TrackStart) -> None:
+        """Handles track start events."""
+        print(f"Track started on guild: {event.guild_id}")
+
+    async def track_finish(self, _: lavasnek_rs.Lavalink, event: lavasnek_rs.TrackFinish) -> None:
+        """Handles track finish events."""
+        print(f"Track finished on guild: {event.guild_id}")
+
+    async def track_exception(
+        self, lavalink: lavasnek_rs.Lavalink, event: lavasnek_rs.TrackException
+    ) -> None:
+        """Handles track exception events."""
+        print(f"Track exception event happened on guild: {event.guild_id}")
+
+        # If a track was unable to be played, skip it
+        skip = await lavalink.skip(event.guild_id)
+        node = await lavalink.get_guild_node(event.guild_id)
+
+        if skip and node:
+            if not node.queue and not node.now_playing:
+                await lavalink.stop(event.guild_id)
+
+
+class Client(tanjun.Client):
+    def __init__(self, *args: typing.Any, **kwargs: typing.Any) -> None:
+        super().__init__(*args, **kwargs)
+        # Tanjun has a metadata property to store any general data
+        # We will store the lavalink client there
+        self.metadata["lavalink"] = None
```

Comment:
Could this not be handled as a type dependency? would be more type safe
Something like
```py
client = Client.from_gateway_bot(...)

@client.with_listener(hikari.ShardReadyEvent)
async def on_ready(event: ..., client_: tanjun.Client = tanjun.injected(type=tanjun.Client)) -> None:
    builder = (
            lavasnek_rs.LavalinkBuilder(event.my_user.id, TOKEN)
            .set_host(os.environ["LAVALINK_HOST"])
            .set_password(os.environ["LAVALINK_PASSWORD"])
            .set_port(int(os.environ["LAVALINK_PORT"]))
            .set_start_gateway(False)
            # We set start gateway False because hikari can handle
            # voice events for us.
    )

    client_.set_type_dependency(lavalink_rs.Lavalink, await builder.build(EventHandler))
```

with this all just negating the need for subclassing as you'd be able to access the client at runtime like this
```py
@client.with_listener(...)/component.with_listener(...)
async def on_voice_state_update(event: ..., lavalink: tanjun.injected(type=lavasnek_rs.Lavalink)) -> None:
    ...

@component.with_slash_command
@tanjun.as_slash_command("name", "description")
async def name_command(ctx: tanjun.abc.Context, lavalink: tanjun.injected(type=lavasnek_rs.Lavalink) -> None:
    ...
```

Reply:
i will take a look at setting it up this way. it does feel much more tanjun-esque.

Reply:
This is now implemented.